source/main/commands/dir-project-export.ts

```diff
@@ -56,44 +61,48 @@ export default class DirProjectExport extends ZettlrCommand {
     // Import our RE
     let fnExportRE = getFnExportRE(true) // We want the multiline version.
 
-    // Concat the files
-    let contents: string[] = []
-    for (let file of files) {
-      const fileDescriptor = await this._app.getFileSystem().getFileContents(file)
-      let fileContents: string = fileDescriptor.content
-      // Directly make all image paths absolute to prevent errors if used
-      // in nested project directories (in which this._dir.path is not the
-      // same for all files). Also make the footnotes unique to prevent
-      // assigning errors.
-      fileContents = makeImgPathsAbsolute(file.dir, fileContents)
+    let promiseArr: Array<Promise<MDFileMeta | CodeFileMeta>> = []
+    files.forEach(element => {
+      promiseArr.push(this._app.getFileSystem().getFileContents(element))
+    })
 
-      fileContents = fileContents.replace(fnExportRE, (match, p1: string, offset, string) => `[^${String(file.hash)}${p1}]`)
+    const contentsArr = await Promise.all(promiseArr)
+    const fileContentArr: string[] = []
 
-      contents.push(fileContents)
-    }
-    const finalContents = contents.join('\n\n')
+    contentsArr.forEach((value, index) => {
+      let fileContent = makeImgPathsAbsolute(path.dirname(files[index]), value.content)
+
+      if (!(contentsArr.length === 1)) {
+        fileContent = fileContent.replace(fnExportRE, (match, p1: string, offset, string) => `[^${String(files[index].hash)}${p1}]`)
+      }
 
+      fileContentArr.push(fileContent.trim() + `${EOL}${EOL}`)
+    })
+
+    // TODO: This only extracts frontmatter at the top of the concatenated file. If there are other 'frontmatter' values in the document they should be extracted too!
+    const splitContent = matter(fileContentArr.join())
```

Comment:
Better: In the FSAL file loader, there is a function `extractYamlFrontmatter` (or similar, but the first word is definitely extract), which basically does exactly what this one does here.

I would suggest to replace the custom extract function with this one, and save it in the descriptors. This way you have the front matter for every file, and the only thing we need to change is change the exporter so that it accepts a "full" descriptor instead of the mock object. Also makes this function smaller and integrates it better with the FSAL! And it saves a lot of computing power since the frontmatter will be in the files at every point!

The only thing you'd need to take care of is that you write the extracted frontmatter into the `frontmatter`-prop of the file (or null, if the frontmatter is empty), so that no renderer code breaks (which expects the frontmatter properties to be properties on the `frontmatter`-property … there's too much property in here)